### PR TITLE
less verbose type mismatch message; fixes https://github.com/nim-lang/RFCs/issues/87

### LINKED ISSUE #87: [RFC] type mismatch error message improvement
- [ ] The following would be clearer and more concise; especially useful when lots of overloads are shown:

```
test.nim(883, 12) Error: type mismatch: got <Stream, TaintedString>
but expected one of:
proc readLine(f: File; line: var TaintedString): bool
  first type mismatch at position: 1
  required type: File
  but expression 'outp' is of type: Stream
proc readLine(f: File): TaintedString
  first type mismatch at position: 1
  required type: File
  but expression 'outp' is of type: Stream

expression: readLine(outp, line)
      if outp.readLine(line):
```
=>
```
test.nim(883, 12) {if outp.readLine(line):} Error: type mismatch: got <Stream, TaintedString> in expression: readLine(outp, line); candidates:
  proc readLine(f: File {Stream}; line: var TaintedString): bool
  proc readLine(f: File {Stream}): TaintedString
```

**EDIT**
- [ ] furthermore, we should show fully qualified types instead of types, to make it easier to find where symbols are defined and remove ambiguities, eg in case Stream is defined in multiple modules:
```
  proc readLine(f: File {Stream}; line: var TaintedString): bool
=>
  proc readLine(f: mod1.File {mod2.Stream}; line: var mod3.TaintedString): bool
```

**EDIT** the overloads should also show the module + file where they came from (not obvious in complex cases)

### PR DESCRIPTION
**Pull request by ringabout**
fixes https://github.com/nim-lang/RFCs/issues/87
implements https://forum.nim-lang.org/t/9514

screenshot

image

### CODE REVIEW

**Comment by Araq** [comment]
Don't have that behind a switch, activate it by default. If tests prove too difficult to patch, introduce a switch for these to turn back the old behavior.

**Comment by chancyk** [comment]
@ringabout Hey, thanks for putting that together! I took a crack at displaying the type information for the expression as well and was able to produce this:

image

So this visual exercise would be scanning the matching `[1]` or `[2]` from each section. I split out the "wanted" type to immediately follow the `[]`which prevents having mentally parse the full type signature, but it may be unnecessarily verbose.

Here it is with my original issue of having the extra 9 overloads from the `vmath` package.:

image

I've cross-posted this back to the forum also for feedback.

@Araq I could take a crack at stripping the more verbose form that's subverted by the `nimConciseTypeMismatch` switch if this format looks desirable. Caveat, I know very little about the internals of Nim, but I was able to build from source and test with the `pr_type_mismatch` PR. I can possibly provide a commit to the PR that mostly works. 😅

File: compiler/semcall.nim [hunk 1]
```diff
@@ -277,6 +282,7 @@ proc presentFailedCandidates(c: PContext, n: PNode, errors: CandidateErrors):
 const
   errTypeMismatch = "type mismatch: got <"
   errButExpected = "but expected one of:"
+  errExpectedPosition = "Expected one of (mismatch at position [#]):"
```
Comment (Menduist) [inline compiler/semcall.nim]:
```suggestion
  errExpectedPosition = "Expected one of (first mismatch at position [#]):"
```

Or show multiple mismatchs, I guess

**Comment by ringabout** [comment]
I don't repeat the parameters. It seems to add some noises.

image

**Comment by ringabout** [comment]
> Don't have that behind a switch, activate it by default. If tests prove too difficult to patch, introduce a switch for these to turn back the old behavior.

Sure

**Comment by ringabout** [comment]
> Don't have that behind a switch, activate it by default. If tests prove too difficult to patch, introduce a switch for these to turn back the old behavior.

The old switch provides extra infos though for the exact error messages though. Probably we need to add extra infos according to verbosity.

File: changelog.md [hunk 2]
```diff
@@ -78,6 +78,8 @@
 
 - Removed the `nimIncrSeqV3` define.
 
+- Added the `nimLegacyTypeMismatch` switch to get legacy type mismatch error messages.
```
Comment (Araq) [inline changelog.md:81]: We don't implement new switches via defines.

Comment (ringabout) [reply changelog.md:81]: I will add it to legacy feature then.

**Comment by ringabout** [comment]
Succeeded by https://github.com/nim-lang/Nim/pull/21191

**Pull request closed without merging**